### PR TITLE
17: Add advanced preference sampling methods

### PR DESCRIPTION
**Pull request by flyingpalli**
Fixes issue #17 

What?

- add methods for sampling preferences

How?

- add methods in `sac_rlhf.py`
- change `replay_buffer.sample_trajectories` to sample two lists of `mb_size` many trajectories that have the same length
- add methods to calculate disagreement and entropy in `reward_net.py` (cf. [pebble implementation](https://github.com/rll-research/BPref/blob/main/reward_model.py#L559))

Ready to merge?

- [x] Linter passes
- [x] PR approved

### CODE REVIEW

**Comment by flyingpalli** [comment]
I am just noticing that the sampling from the replay buffer is still far from perfect:
Notice, that each row of `rb.dones` consists of `num_envs` many entries.

Previously, we used to let sampled trajectories be a sequence of rows where only the first and the last row could have 1 as an entry in `dones`. This happend without taking into consideration, in which column (i.e. environment) the `done` occured and let to trajectories being very short.

With the present implementation, we let trajectories be sequences _of fixed length_ which are determined by only applying the previous condition to one column (i.e. there exists one column where the trajectory is the beginning of an observation after a reset). This allows for having `2*mb_size` many trajectories with the same length.

Unfortunately though, now there can be columns in the trajectory where the observation is interrupted by a reset.

I made some runs on 16/12/2024 which can be found in this project's [wandb](https://wandb.ai/cleanRLHF) and I find the progress in the first 40k steps to be sufficient but I can't explain the sharp and irretrievable drop in return after ~45k steps, which is why I write this comment.

File: framework/sac_rlhf.py [hunk 1]
```diff
@@ -253,7 +255,40 @@ def load_replay_buffer(replay_buffer: ReplayBuffer, path: str):
...
+
+
+def entropy_sampling(rb: ReplayBuffer, reward_net: RewardNet):
+    traj_mb1, traj_mb2 = rb.sample_trajectories()
+
+    entropies = []
+    for traj_1, traj_2 in zip(traj_mb1, traj_mb2):
+        probs = []
+        for member in range(len(reward_net.ensemble)):
+            probs.append(reward_net.preference_prob_hat_member(traj_1, traj_2, member))
```
Comment (bkuen) [inline framework/sac_rlhf.py:287]: This must be `reward_net.preference_hat_entropy_member(traj_1, traj_2, member)`,

File: framework/sac_rlhf.py [hunk 2]
```diff
@@ -253,7 +255,40 @@ def load_replay_buffer(replay_buffer: ReplayBuffer, path: str):
     replay_buffer.ground_truth_rewards = buffer_data["ground_truth_rewards"]
     replay_buffer.dones = buffer_data["dones"]
 
-    logging.debug(f"Replay buffer loaded from {path}")
+    logging.info(f"Replay buffer loaded from {path}")
+
+
+def uniform_sampling(rb: ReplayBuffer, reward_net: RewardNet):
```
Comment (bkuen) [inline framework/sac_rlhf.py:261]: Consider moving those three files into a seperate file (e.g. sampling.py) to keep the main file clean.

File: framework/sac_rlhf.py [hunk 3]
```diff
@@ -253,7 +255,40 @@ def load_replay_buffer(replay_buffer: ReplayBuffer, path: str):
...
+    traj_mb1, traj_mb2 = rb.sample_trajectories()
+    return traj_mb1[0], traj_mb2[0]
+
+
+def disagreement_sampling(rb: ReplayBuffer, reward_net: RewardNet):
+    traj_mb1, traj_mb2 = rb.sample_trajectories()
+
+    disagrees = []
```
Comment (bkuen) [inline framework/sac_rlhf.py:269]: Add `with torch.no_grad():` before for loop to avoid unnecessary calculation of gradients

File: framework/sac_rlhf.py [hunk 4]
```diff
@@ -253,7 +255,40 @@ def load_replay_buffer(replay_buffer: ReplayBuffer, path: str):
...
+        disagrees.append(np.std(probs, axis=0))
+
+    disagrees_argmax = max(enumerate(disagrees), key=lambda x: x[1])[0]
+    return traj_mb1[disagrees_argmax], traj_mb2[disagrees_argmax]
+
+
+def entropy_sampling(rb: ReplayBuffer, reward_net: RewardNet):
+    traj_mb1, traj_mb2 = rb.sample_trajectories()
+
+    entropies = []
```
Comment (bkuen) [inline framework/sac_rlhf.py:283]: Add `with torch.no_grad():` before for loop to avoid unnecessary calculation of gradients

**Review by bkuen — CHANGES_REQUESTED** [review]
In general, this seems to be a fine first implementation of sampling. It still offers potential for improvement. I am sure that some for loops could be replaced by larger tensor operations. But let's do that in a separate issue if we experience any preformance bottlenecks.

**Review by bkuen — DISMISSED** [review]
Thank you for adjusting the code as suggested. As this adds a lot of additional complexity, I suggest that someone else also takes a look at the code before merging.

**Review by jualat — APPROVED** [review]
Looks good to me. The sampling strategies look like they do what they are supposed to do and the code works as expected for me. Nice work :+1: 
The change of the logging level is also fine.

**Review by jualat — CHANGES_REQUESTED** [review]
Just a small thing: In `sampling.py` there is the import statement for numpy missing

**Comment by jualat** [comment]
> Just a small thing: In `sampling.py` there is the import statement for numpy missing

After correcting this the code was running fine for me

**Review by jualat — APPROVED** [review]
